src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3476,29 +3478,45 @@ protected override void VisitStatement(BoundStatement statement)
             // https://github.com/dotnet/roslyn/issues/68786: Use inferInitialObjectState() to set the initial
             // state of the instance: see the call to InheritNullableStateOfTrackableStruct() in particular.
             int containerSlot = GetOrCreatePlaceholderSlot(node);
-            bool delayCompletionForType = false; // https://github.com/dotnet/roslyn/issues/68786: Should be true if the collection expression is target typed.
+            NullableFlowState resultState = NullableFlowState.NotNull;
+
+            var collectionKind = ConversionsBase.GetCollectionExpressionTypeKind(this.compilation, node.Type, out var targetElementType);
+            if (collectionKind is CollectionExpressionTypeKind.CollectionBuilder)
+            {
+                var createMethod = node.CollectionBuilderMethod;
+                if (createMethod is not null)
+                {
+                    var readOnlySpan = (NamedTypeSymbol)createMethod.Parameters[0].Type;
+                    targetElementType = readOnlySpan.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                    var annotations = createMethod.GetFlowAnalysisAnnotations();
+                    resultState = ApplyUnconditionalAnnotations(createMethod.ReturnTypeWithAnnotations, annotations).ToTypeWithState().State;
+                }
+            }
 
-            // https://github.com/dotnet/roslyn/issues/68786: Set nullability of elements from the inferred target type nullability.
             foreach (var element in node.Elements)
             {
                 switch (element)
                 {
                     case BoundCollectionElementInitializer initializer:
                         var collectionType = initializer.AddMethod.ContainingType;
-                        var completion = VisitCollectionElementInitializer(initializer, collectionType, delayCompletionForType);
-                        if (completion is { })
-                        {
-                            // https://github.com/dotnet/roslyn/issues/68786: Complete the analysis later.
-                            completion(containerSlot, collectionType);
-                        }
+
+                        var completion = VisitCollectionElementInitializer(initializer, collectionType,
+                            delayCompletionForType: false /* All collection expressions are target-typed */);
+
+                        Debug.Assert(completion is null);
+                        break;
+                    case BoundCollectionExpressionSpreadElement spread:
+                        // https://github.com/dotnet/roslyn/issues/68786: We should check the spread
+                        Visit(spread);
```

Comment:
I think we may want VisitRvalue here because if a spread puts us into a split state after visiting (like in an error scenario) that could mess things up later on.

Reply:
FWIW, with https://github.com/dotnet/roslyn/pull/70197, `NullableWalker` relies on `base.VisitCollectionExpressionSpreadElement(node)` which simply calls `VisitRvalue(node.Expression)`.

Reply:
I'd be comfortable with this assuming we have a test demonstrating that things don't fall over when an expression inside a spread splits the state. e.g. `[..(x is not null)]`

Reply:
Spreads are not in scope for this PR, so I don't want to go too far. I'd be okay with removing this entirely if you prefer.

As things stand, `..(x is null)` hits an assertion: `System.InvalidOperationException : Analyzed 11 nodes in NullableWalker, but DebugVerifier expects 10. Example of unverified node: BoundCollectionExpressionSpreadElement ..(x is null)`

In release mode, the test produces the expected diagnsotic (binding error). Added a test

Reply:
Note: a similar assertion is hit if I remove this case. I still think it's beneficial to keep it in the interim (before we have fully coverage on spreads)

Reply:
As long as the compiler doesn't blow up in the retail build in these cases, I'm happy. Thanks!

Reply:
Chuck pointed out the assertion problem can be solved by undoing the `s_skippedExpressions` change. Done